Launch simple MongoDB replica set

njhale opened the pull request:
Configure the MongoDB deployment launched by this Acorn to run as
a MongoDB replica set of size 1, using keyfiles for member authentication.
Running as a replica set enables MongoDB transactions and will
allow clients that depend on them to consume this Acorn; e.g. Prisma's
MongoDB database connector.

Addresses #6

njhale committed:
Launch simple MongoDB replica set

Configure the MongoDB deployment launched by this Acorn to run as
a MongoDB replica set of size 1, using keyfiles for member authentication.
Running as a replica set enables MongoDB transactions and will
allow clients that depend on them to consume this Acorn; e.g. Prisma's
MongoDB database connector.

Signed-off-by: Nick Hale <4175918+njhale@users.noreply.github.com>

njhale commented on Acornfile:
```diff
@@ -5,100 +5,127 @@ info:        localData.info
+					// Replica set keys must be between 6 and 1024 characters in length and consist of only valid base64 characters.
+					// See https://www.mongodb.com/docs/v6.2/tutorial/deploy-replica-set-with-keyfile-access-control/#create-a-keyfile
+					// for details.
+					entrypoint: ["bash", "-c", "openssl rand -base64 756 > replica-set-key && chmod 400 replica-set-key && chown 999:999 replica-set-key && mv replica-set-key /var/keys/replica-set-key"]
```
If possible think this sidecar should be simplified to a generated secret that gets mounted as a file w/ the correct permissions (the default permissions for `files` is too permissive for mongo, and it complains)

njhale commented on Acornfile:
```diff
@@ -5,100 +5,127 @@ info:        localData.info
 services: db: {
 	default:   true
-	container: "mongodb"
+	container: "mongodb-0"
```
Since `containerLabels` are disabled in manager for security reasons, we'll need support for setting multiple containers by name before `args.replicas > 1` will work properly.

At this point, I'm also not sure if exposing a mongodb replica set via a service LB will work properly. I've seen a lot of examples of clients consuming connection strings that contain all members.

njhale replied:
see https://github.com/acorn-io/runtime/issues/2327

cloudnautique commented:
For this PR, can you also add a new example in the folder? Will the old example still work, I would expect yes?

njhale commented:
> For this PR, can you also add a new example in the folder? Will the old example still work, I would expect yes?

@cloudnautique Yes, the old example still works and I've removed `args.replicas` from this PR and modified things s.t. there will only ever be a single node deployed, so the usage will remain the same.

njhale commented:
just tested again and this is no longer working after my latest changes, I want to hold off on merging for the moment.

njhale commented on Acornfile:
```diff
@@ -20,85 +20,96 @@ services: db: {
+		// Replica set keys must be between 6 and 1024 characters in length and consist of only valid base64 characters.
+		// See https://www.mongodb.com/docs/v6.2/tutorial/deploy-replica-set-with-keyfile-access-control/#create-a-keyfile
+		// for details.
+		entrypoint: ["bash", "-c", "cd /var/keys && openssl rand -base64 756 > replica-set-key && chmod 400 replica-set-key && chown 999:999 replica-set-key"]
```
Spoke with @cloudnautique offline; going to change this to a generated secret.

njhale replied:
ended up going a slightly different, hopefully simpler, direction with this.

njhale commented on Acornfile:
```diff
@@ -20,84 +20,92 @@ services: db: {
+		}
+	}
 
-localData: readinessProbeCommand: [
```
this wasn't being used.

njhale commented on Acornfile:
```diff
@@ -20,84 +20,92 @@ services: db: {
-        }
-    }
-}
+	"replica-set-key": type: "token"
```
I ended up _not_ creating a generated secret and base64 encoding this token (a mongo requirement) in the new startup script. The result is one less job and a cleaner Acornfile.

njhale commented on Acornfile:
```diff
@@ -20,84 +20,92 @@ services: db: {
+		}
+		dependsOn: ["mongodb"]
+		entrypoint: ["/acorn/scripts/render.sh"]
+		dirs: "/acorn/scripts": "./scripts"
```
this lets us drop the custom Dockerfile that was used for this job

njhale commented on scripts/start-mongodb.sh:
```diff
@@ -0,0 +1,11 @@
+# for details.
+echo -n "${REPLICA_SET_KEY}" | base64 > /replica-set-key
+chmod 0400 /replica-set-key
+chown 999:999 /replica-set-key
```
this is required for mongod to use the file

njhale commented on Acornfile:
```diff
@@ -20,84 +20,92 @@ services: db: {
+	env: {
+		MONGO_INITDB_ROOT_USERNAME: "secret://admin/username"
+		MONGO_INITDB_ROOT_PASSWORD: "secret://admin/password"
+		"REPLICA_SET_KEY":          "secret://replica-set-key/token"
```
It would be nice to mount the content of a generated secret as a file, but mongod requires that we run `chmod 999:999` on the file before it will read it.

njhale reviewed (commented):
OK, this is working and ready for review
